dune_items.ml

```diff
@@ -0,0 +1,201 @@
+  let bin_name s =
+    Astring.String.cut ~sep:"/" ~rev:true s
+    |> Option.map snd |> Option.get
+
+  (* With "defautl/lib/bin.exe", it gives "default/lib" *)
+  let source_dir s =
+    Astring.String.cut ~sep:"/" ~rev:true s
+    |> Option.map fst |> Option.map (Astring.String.cut ~sep:"/" ~rev:false) |> Option.join |> Option.map snd
+    |> function None -> "." | Some dir -> dir
+
+  let decode_item sexps =
+    List.fold_left (fun item sexps ->
+        match sexps with
+        | Sexp.List [Atom "src"; List [_; Atom p] ] ->
+          {item with source_dir = source_dir p; bin_name = bin_name p}
+        | Sexp.List [Atom "kind"; Atom p ]    -> {item with kind = p}
+        | Sexp.List [Atom "dst"; Atom p ]     -> {item with dst = p}
+        | Sexp.List [Atom "section"; Atom p ] -> {item with section = p}
+        | Sexp.List [Atom "optional"; Atom p ] -> {item with optional = p}
+        | s -> Fmt.failwith "%s is not a good format decoding an item" (Sexp.to_string s)
+      ) dump_item sexps
+    |> (fun item -> match item.section with "BIN" -> Bin item | _ -> Other item)
+
+  let decode_items : Sexp.t list -> entry list =
+    List.filter_map (function
+        | Sexp.List [List [Atom "source"; List [Atom "User" ; _ ]]; List [Atom "entry"; List sexps]] -> Some (decode_item sexps)
+        | Sexp.List [List [Atom "entry"; List sexps]; List [Atom "source"; Atom "User"]] -> Some (decode_item sexps)
+        | Sexp.List [List [Atom "source"; Atom "Dune"]; List _ ] -> None
+        | s -> Fmt.failwith "%s is not a good format decoding items" (Sexp.to_string s))
+
+  let decode_entries : Sexp.t -> t = function
+    | Sexp.List [Atom package; List sexps] -> (package,decode_items sexps)
+    | _ -> Fmt.failwith "Invalid format"
+
+  let entries_of_sexp : Sexp.t -> t list = function
+    | Sexp.List sexps ->
+      List.map decode_entries sexps
+      |> List.map (fun (package, entries) ->
+          (package, List.map (function
+               | Bin item   -> Bin {item with package = package}
+               | Other item -> Other {item with package = package}) entries))
+    | _ -> Fmt.failwith "Invalid format"
+
+  let items_bin_of_entries describe_entries =
+    List.concat_map snd describe_entries
+    |> List.filter_map (fun d_item ->
+        d_item
+        |> (function Bin item -> Some (item.bin_name,item) | Other _ -> None))
+    |> List.to_seq |> Item_map.of_seq
+end
+
+module Describe_opam_files = struct
+
+  type t = (string * OpamFile.OPAM.t ) list
+
+  let decode_items = function
+    | Sexp.List sexps ->
+      sexps
+      |> List.map (function
+          | Sexp.List [Atom opam_file; Atom opam_content] ->
+            (opam_file, OpamFile.OPAM.read_from_string opam_content)
+          | s -> Fmt.failwith "%s is not a good format decoding an item" (Sexp.to_string s))
+    | s -> Fmt.failwith "%s is not a good format decoding items" (Sexp.to_string s)
+
+  let opam_files_of_sexp = decode_items
```

Comment:
What about this with comments on the types and an explanation of the opam_files_of_sexp function?
```suggestion
   (** String representing an opam_file name eg. foo.opam *)
  type opam_file = string

  (** ??? *)
  type t = (opam_file * OpamFile.OPAM.t ) list

  (** Decode opam_files from dune describe output. *)
  let opam_files_of_sexp = function
    | Sexp.List sexps ->
      sexps
      |> List.map (function
          | Sexp.List [Atom opam_file; Atom opam_content] ->
            (opam_file, OpamFile.OPAM.read_from_string opam_content)
          | s -> Fmt.failwith "%s is not a good format decoding an item" (Sexp.to_string s))
    | s -> Fmt.failwith "%s is not a good format decoding items" (Sexp.to_string s)
```